machineconfiguration/v1/types_machineosconfig.go

```diff
@@ -0,0 +1,228 @@
+// +union
+type BuildOutputs struct {
+	// currentImagePullSecret is the secret used to pull the final produced image.
+	// must live in the openshift-machine-config-operator namespace
+	// the final image push and pull secrets should be separate for security concerns. If the final image push secret is somehow exfiltrated,
+	// that gives someone the power to push images to the image repository. By comparison, if the final image pull secret gets exfiltrated,
+	// that only gives someone to pull images from the image repository. It's basically the principle of least permissions.
+	// this pull secret will be used on all nodes in the pool. These nodes will need to pull the final OS image and boot into it using rpm-ostree or bootc.
+	// +optional
+	CurrentImagePullSecret ImageSecretObjectReference `json:"currentImagePullSecret,omitempty"`
+}
+
+type MachineOSImageBuilder struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// +kubebuilder:default:=PodImageBuilder
+	// +kubebuilder:validation:Enum:=PodImageBuilder
+	// Valid options are: PodImageBuilder
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+}
+
+// MachineOSContainerfile contains all custom content the user wants built into the image
+type MachineOSContainerfile struct {
+	// containerfileArch describes the architecture this containerfile is to be built for
+	// this arch is optional. If the user does not specify an architecture, it is assumed
+	// that the content can be applied to all architectures, or in a single arch cluster: the only architecture.
+	// +kubebuilder:validation:Enum:=arm64;amd64;ppc64le;s390x;aarch64;x86_64;noarch
+	// +kubebuilder:default:=noarch
+	// +optional
+	ContainerfileArch ContainerfileArch `json:"containerfileArch"`
+	// content is the custom content to be built
+	// +kubebuilder:validation:Required
+	Content string `json:"content"`
+}
+
+type ContainerfileArch string
+
+const (
+	// describes the arm64 architecture
+	Arm64 ContainerfileArch = "arm64"
+	// describes the amd64 architecture
+	Amd64 ContainerfileArch = "amd64"
+	// describes the ppc64le architecture
+	Ppc ContainerfileArch = "ppc64le"
+	// describes the s390x architecture
+	S390 ContainerfileArch = "s390x"
+	// describes the aarch64 architecture
+	Aarch64 ContainerfileArch = "aarch64"
+	// describes the fx86_64 architecture
+	X86_64 ContainerfileArch = "x86_64"
+	// describes a containerfile that can be applied to any arch
+	NoArch ContainerfileArch = "noarch"
+)
+
+// Refers to the name of a MachineConfigPool (e.g., "worker", "infra", etc.):
+// the MachineOSBuilder pod validates that the user has provided a valid pool
+type MachineConfigPoolReference struct {
+	// name of the MachineConfigPool object.
+	// +kubebuilder:validation:MaxLength:=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
```

Comment:
We can now use the CEL format validations for this

Reply:
Same comment as above, temporarily reverted until we can get it working